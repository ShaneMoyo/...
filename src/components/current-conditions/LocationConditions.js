import React, { Component } from 'react';
import superagent from 'superagent';
import PropTypes from 'prop-types';

const API_KEY = process.env.REACT_APP_API_KEY || null;

class LocationConditions extends Component {

  state = { loading: false, validationFailed: false };

  componentDidMount(){
    const { id } = this.props;
    const storedZip = localStorage.getItem(id);
    if(storedZip) this.getConditions(storedZip);
  }

  checkZip = zipCode => {
    const regex = /^\d{5}$/;
    if(!regex.test(zipCode)){
      return false;
    }
    if(zipCode === ' '){
      return false;
    }
    return true;
  }

  handleSubmit = event => {
    event.preventDefault();
    const { value: zipCode } = event.target.elements.zip;
    if(this.checkZip(zipCode)) {
      this.setState({ validationFailed: false });
      return this.getConditions(zipCode);
    }
    this.setState({ validationFailed: true });
  };

  getConditions = zipCode => {
    this.setState({ loading: true });
    const url = `http://api.wunderground.com/api/${API_KEY}/conditions/q/${zipCode}.json`;
    const { id, handleGotConditions } = this.props;

    return superagent.get(url) 
      .then(({ body: gotConditions }) => {
        if(gotConditions.response.error) {
          const { description: error } =  gotConditions.response.error;
          gotConditions = { id, error };
        } else {
          const { weather, temp_f: temperature, icon, display_location } = gotConditions.current_observation;
          gotConditions = { id, weather, temperature, icon, location: display_location.full };
          localStorage.setItem(id, zipCode);
        }
        handleGotConditions(gotConditions);
        this.setState({ loading: false });
      });
  }

  render(){
    const { weather, temperature, error, icon, location } = this.props.conditions;
    const { loading, validationFailed } = this.state;

    const buttonStyle = loading ? 'button is-loading is-small is-info' : 'button is-outlined is-small is-info';
    const inputStyle = validationFailed ? 'input is-danger is-small' : 'input is-small';

    const view = !error ? 
      <div>
        <h1 className="title is-4 has-text-grey">{location}</h1>
        { icon ? <img src={icon}  alt="weather status icon"/> : null}
        <h1 className="title is-4 has-text-grey">{weather}</h1>
        <h1 className="subtitle has-text-info">{temperature}</h1>
        <br/>
      </div> : 
      <div>
        <h1 className="title is-4 has-text-grey">{error}</h1>
      </div>;

    return(
      <div className="column is-one-third has-text-centered">
        <div className="box is-offset-2"> 
          {view}
          <form onSubmit={event => this.handleSubmit(event)}>
            <div className="field">
              <div className="control">
                <input className={inputStyle} type="text" placeholder="enter zip-code" name="zip"/>
                { validationFailed ? <p class="help is-danger">The lenght of zip code must be five digits</p> : null}
              </div>
            </div>
            <button className={buttonStyle} type="submit">Submit</button>
          </form>
        </div>
      </div>
    );
  }
}

LocationConditions.propTypes = {
  conditions: PropTypes.object,
  id: PropTypes.string,
  handleGotConditions: PropTypes.func
};

export default LocationConditions;