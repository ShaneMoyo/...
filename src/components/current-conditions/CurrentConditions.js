import React, { Component } from 'react';
import LocationConditions from './LocationConditions';

class CurrentConditions extends Component {

  state = {
    locationA: {
      location: null,
      weather: null,
      temperature: null,
      icon: null,
      error: null,
    },
    locationB: {
      location: null,
      weather: null,
      temperature: null,
      icon: null,
      error: null,
    }
  };

  handleLoadConditions = gotConditions => {
    const { id, weather, temperature, icon, location, error } = gotConditions;
    const newState = { ...this.state };
    newState[id] = !error ? { weather, temperature, location, icon: `https://icons.wxug.com/i/c/i/${icon}.gif` } : { error };
    this.setState(newState);
    this.calculateTemperatureDifference();
  };

  calculateTemperatureDifference = () => {
    const { locationA, locationB } = this.state;
    if(!locationA.temperature || !locationB.temperature) return;
    const temperatureDifference = Math.round((locationA.temperature - locationB.temperature) * 100) / 100;
    this.setState({ temperatureDifference });
  };

  render() {
    const { locationA, locationB, temperatureDifference } = this.state;
    const showTemperatureDifference = locationA.temperature && locationB.temperature;

    return (
      <div className="container is-fluid animated fadeIn" >
        <div className="columns is-centered">

          <LocationConditions id="locationA"
            conditions={locationA} 
            handleLoadConditions={response => this.handleLoadConditions(response)}/>
                
          <LocationConditions id="locationB"
            conditions={locationB}
            handleLoadConditions={response => this.handleLoadConditions(response)}/>

        </div>

        { showTemperatureDifference ? 
          <div className="container has-text-centered">
            <h1 className="title is-4 has-text-grey">
              {`Temperature difference between ${locationA.location} and ${locationB.location}: `}
            </h1>
            <p className="subtitle has-text-info" >{`${temperatureDifference} degrees`}</p>
          </div> : null }
          
      </div>
    );
  }
}

export default CurrentConditions;
