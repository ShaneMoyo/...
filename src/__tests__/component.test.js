import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import CurrentConditions from '../components/current-conditions/CurrentConditions';
import LocationConditions from '../components/current-conditions/LocationConditions';

const conditions = { 
  location: 'test',
  weather: 'test',
  temperature: 70,
  icon: null,
  error: null
};

describe('component snapshot tests', () => {

  it('should render CurrentConditions component', () => {
    const wrapper = shallow(<CurrentConditions/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('should render LocationConditions component', () => {
    const wrapper = shallow(<LocationConditions conditions={conditions}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

});