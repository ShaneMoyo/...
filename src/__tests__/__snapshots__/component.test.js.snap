// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component snapshot tests should render CurrentConditions component 1`] = `
<div
  class="container is-fluid"
>
  <div
    class="columns is-centered"
  >
    <LocationConditions
      conditions={
        Object {
          "error": null,
          "icon": null,
          "location": null,
          "temperature": null,
          "weather": null,
        }
      }
      handleLoadConditions={[Function]}
      id="location1"
    />
    <LocationConditions
      conditions={
        Object {
          "error": null,
          "icon": null,
          "location": null,
          "temperature": null,
          "weather": null,
        }
      }
      handleLoadConditions={[Function]}
      id="location2"
    />
  </div>
</div>
`;

exports[`component snapshot tests should render LocationConditions component 1`] = `
<div
  class="column is-one-third has-text-centered"
>
  <div
    class="box is-offset-2"
  >
    <div>
      <h1
        class="title is-4 has-text-grey"
      >
        test
      </h1>
      <h1
        class="title is-4 has-text-grey"
      >
        test
      </h1>
      <h1
        class="subtitle has-text-info"
      >
        70
      </h1>
      <br />
    </div>
    <form
      onSubmit={[Function]}
    >
      <div
        class="field"
      >
        <div
          class="control"
        >
          <input
            class="input is-small"
            name="zip"
            placeholder="enter zip-code"
            type="text"
          />
        </div>
      </div>
      <button
        class="button is-outlined is-small is-info"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
